internal/k8s/reconciler/gateway.go

```diff
@@ -77,258 +72,7 @@ func newK8sGateway(gateway *gwv1beta1.Gateway, config K8sGatewayConfig) *K8sGate
-// This IP address is not always externally accessible and may require additional
-// work by the practitioner such as port-forwarding or opening firewall rules to make
-// it externally accessible.
-func (g *K8sGateway) assignGatewayIPFromPodHost(ctx context.Context) error {
-	pods, err := g.client.PodsWithLabels(ctx, utils.LabelsForGateway(g.Gateway))
-	if err != nil {
-		return err
-	}
-
-	if len(pods) == 0 {
-		g.GatewayState.Status.Scheduled.NotReconciled = errors.New("pods not found")
-		return nil
-	}
-
-	for _, pod := range pods {
-		if pod.Status.HostIP != "" {
-			g.GatewayState.ServiceReady = true
-
-			if !slices.Contains(g.GatewayState.Addresses, pod.Status.HostIP) {
-				g.GatewayState.Addresses = append(g.GatewayState.Addresses, pod.Status.HostIP)
-			}
-		}
-	}
-
-	return nil
-}
-
-func (g *K8sGateway) validatePods(ctx context.Context) error {
-	pods, err := g.client.PodsWithLabels(ctx, utils.LabelsForGateway(g.Gateway))
-	if err != nil {
-		return err
-	}
-
-	for _, pod := range pods {
-		g.validatePodConditions(&pod)
-	}
-
-	return nil
-}
-
-func (g *K8sGateway) validatePodConditions(pod *corev1.Pod) {
-	if pod == nil {
-		g.GatewayState.Status.Scheduled.NotReconciled = errors.New("pod not found")
-		return
-	}
-
-	switch pod.Status.Phase {
-	case corev1.PodPending:
-		g.validatePodStatusPending(pod)
-	case corev1.PodRunning:
-		g.validatePodStatusRunning(pod)
-	case corev1.PodSucceeded:
-		// this should never happen, occurs when the pod terminates
-		// with a 0 status code, consider this a failed deployment
-		fallthrough
-	case corev1.PodFailed:
-		// we have a failed deployment, set the status accordingly
-		// for now we just consider the pods unschedulable.
-		g.GatewayState.Status.Scheduled.PodFailed = errors.New("pod not running")
-	default: // Unknown pod status
-		// we don't have a known pod status, just consider this unreconciled
-		g.GatewayState.Status.Scheduled.Unknown = errors.New("pod status unknown")
-	}
-}
-
-func (g *K8sGateway) validatePodStatusPending(pod *corev1.Pod) {
-	for _, condition := range pod.Status.Conditions {
-		if condition.Type == corev1.PodScheduled && condition.Status == corev1.ConditionFalse &&
-			strings.Contains(condition.Reason, "Unschedulable") {
-			g.GatewayState.Status.Scheduled.NoResources = errors.New(condition.Message)
-			return
-		}
-	}
-	// if no conditions exist, or we haven't found a specific above condition, just default
-	// to not reconciled
-	g.GatewayState.Status.Scheduled.NotReconciled = errors.New("pod conditions not found")
-}
-
-func (g *K8sGateway) validatePodStatusRunning(pod *corev1.Pod) {
-	for _, condition := range pod.Status.Conditions {
-		if condition.Type == corev1.PodReady && condition.Status == corev1.ConditionTrue {
-			g.GatewayState.PodReady = true
-			return
-		}
-	}
+	return g.validator.Validate(ctx, g.GatewayState, g.Gateway, g.deployer.Service(g.config, g.Gateway))
```

Comment:
Wondering, did you keep the validator as part of the gateway to reduce the test changes that would need to be introduced to strip out `Validate` from the `K8sGateway` struct? I believe that the original PR had this call in the Manager? Is there a plan to move this there eventually?

Reply:
You are correct, and no reason we can't move it out to the manager in a future PR. This PR is already more chunky than I'd like though so tried to narrow as much as possible.